tools/clang/test/HLSLFileCheck/passes/hl/sroa_hlsl/memcpy_preuser.hlsl

```diff
@@ -34,9 +36,9 @@ cbuffer cbuf : register(b1)
 float main(int doit : A) : SV_Target
 {
   float res = 0.0;
+  OuterStruct oStruct;
```

Comment:
@pow2clk I'm slightly confused by the terminology "change" and "alteration" that you use, so apologies if I misunderstood you.
The original code could be replaced entirely by returning undef since the += always adds an uninitialized value. If lifetime intrinsics do not exist, the definition of oStruct is effectively hoisted outside the loop and suddenly the access is not 'undef' anymore but refers to the struct that's copied in the previous iteration of the loop. Of course that transformation is legal since undef can be replaced by an arbitrary value created from non-crashing, terminating code ;).

Now what's happening WITH lifetime intrinsics is not what I describe above: The opt pipeline actually fails to realize that there is an add with undef that it can use to further simplify the code. So a lot of stuff remains but the code looks different from before (and is likely to change if the opt pipeline becomes more aggressive with undef values). The resulting undef usage validation error stems from this, but I'd say that's more of a coincidence.

I believe that the test code should rather be modified to not exercise how much we can optimize undef values but instead do what was obviously originally intended with the i%2 == 0 conditional that refers to the previous iteration.

Reply:
For reference, here's the resulting DXIL of the modified test with lifetime intrinsics (can't attach a file :( ):
; ModuleID = 'D:\msft\DirectXShaderCompiler\tools\clang\test\HLSLFileCheck\passes\hl\sroa_hlsl\memcpy_preuser.hlsl'
target datalayout = "e-m:e-p:32:32-i1:32-i8:32-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64"
target triple = "dxil-ms-dx"

%struct.OuterStruct = type { float, float }
%cbuf = type { [1 x %struct.OuterStruct] }

define void @main() {
entry:
  %0 = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef)
  %cmp.1 = icmp sgt i32 %0, 0
  br i1 %cmp.1, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %res.0 = phi float [ 0.000000e+00, %entry ], [ %add.res.1, %for.cond.cleanup.loopexit ]
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float %res.0)
  ret void

for.body:                                         ; preds = %for.body.preheader, %for.body
  %res.1 = phi float [ %add.res.1, %for.body ], [ 0.000000e+00, %for.body.preheader ]
  %i.0 = phi i32 [ %inc, %for.body ], [ 0, %for.body.preheader ]
  %rem2 = and i32 %i.0, 1
  %cmp2 = icmp eq i32 %rem2, 0
  %add = fadd fast float %res.1, undef
  %add.res.1 = select i1 %cmp2, float %add, float %res.1
  %inc = add nuw nsw i32 %i.0, 1
  %exitcond = icmp eq i32 %inc, %0
  br i1 %exitcond, label %for.cond.cleanup.loopexit, label %for.body
}

; Function Attrs: nounwind readnone
declare i32 @dx.op.loadInput.i32(i32, i32, i32, i8, i32) #0

; Function Attrs: nounwind
declare void @dx.op.storeOutput.f32(i32, i32, i32, i8, float) #1

attributes #0 = { nounwind readnone }
attributes #1 = { nounwind }

!llvm.ident = !{!0}
!dx.version = !{!1}
!dx.valver = !{!2}
!dx.shaderModel = !{!3}
!dx.typeAnnotations = !{!4, !10}
!dx.viewIdState = !{!14}
!dx.entryPoints = !{!15}

!0 = !{!"clang version 3.7 (tags/RELEASE_370/final)"}
!1 = !{i32 1, i32 0}
!2 = !{i32 1, i32 6}
!3 = !{!"ps", i32 6, i32 0}
!4 = !{i32 0, %struct.OuterStruct undef, !5, %cbuf undef, !8}
!5 = !{i32 8, !6, !7}
!6 = !{i32 6, !"fval", i32 3, i32 0, i32 7, i32 9}
!7 = !{i32 6, !"fval2", i32 3, i32 4, i32 7, i32 9}
!8 = !{i32 8, !9}
!9 = !{i32 6, !"g_oStruct", i32 3, i32 0}
!10 = !{i32 1, void ()* @main, !11}
!11 = !{!12}
!12 = !{i32 0, !13, !13}
!13 = !{}
!14 = !{[3 x i32] [i32 1, i32 1, i32 1]}
!15 = !{void ()* @main, !"main", !16, null, null}
!16 = !{!17, !21, null}
!17 = !{!18}
!18 = !{i32 0, !"A", i8 4, i8 0, !19, i8 1, i32 1, i8 1, i32 0, i8 0, !20}
!19 = !{i32 0}
!20 = !{i32 3, i32 1}
!21 = !{!22}
!22 = !{i32 0, !"SV_Target", i8 9, i8 16, !19, i8 0, i32 1, i8 1, i32 0, i8 0, !20}

Reply:
And here's the DXIL of the modified test with lifetime intrinsics, which is basically identical to the original test without lifetimes:

%cbuf = type { [1 x %struct.OuterStruct] }
%struct.OuterStruct = type { float, float }
%dx.types.Handle = type { i8* }
%dx.types.CBufRet.f32 = type { float, float, float, float }

@cbuf = external constant %cbuf

define void @main() {
entry:
  %cbuf_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 1, i1 false)
  %0 = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef)
  %cmp.1 = icmp sgt i32 %0, 0
  br i1 %cmp.1, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %res.0 = phi float [ 0.000000e+00, %entry ], [ %res.2, %for.cond.cleanup.loopexit ]
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float %res.0)
  ret void

for.body:                                         ; preds = %for.body.preheader, %for.body
  %res.1 = phi float [ %res.2, %for.body ], [ 0.000000e+00, %for.body.preheader ]
  %oStruct.1.0 = phi float [ %2, %for.body ], [ undef, %for.body.preheader ]
  %i.0 = phi i32 [ %inc, %for.body ], [ 0, %for.body.preheader ]
  %rem2 = and i32 %i.0, 1
  %cmp2 = icmp eq i32 %rem2, 0
  %add = fadd fast float %oStruct.1.0, %res.1
  %res.2 = select i1 %cmp2, float %add, float %res.1
  %1 = call %dx.types.CBufRet.f32 @dx.op.cbufferLoadLegacy.f32(i32 59, %dx.types.Handle %cbuf_cbuffer, i32 %0)
  %2 = extractvalue %dx.types.CBufRet.f32 %1, 1
  %inc = add nuw nsw i32 %i.0, 1
  %exitcond = icmp eq i32 %inc, %0
  br i1 %exitcond, label %for.cond.cleanup.loopexit, label %for.body
}

Reply:
I recommend you use the three backticks "```" to quote code as it will annotate it and prevent accidentally mentioning github users as has happened here. I don't know why you were unable to attach files. That's a feature github should support.

I edited my original message to try to clarify that I'd performed the tests that you copied here already and what the outcome was.

The reason for the PR this test was made to validate was not to "optimize undef values", but to 1) allow memcpy replacements in circumstances other than this which would result in better optimized code and 2) avoid erroneous behavior in this case and possibly others similar to it we failed to conceive of. This test is as much to ensure that the previous behavior not be mistakenly restored as anything else.

I would like to know two things:
1) Why this PR causes an undef error with the original form of this hlsl file that wasn't produced before.
2) Why the additional CHECK lines were added as they are not newly produced with this PR and have no bearing on what this test is meant to evaluate.

Reply:
Ugh, very good point. Sorry for mentioning other people :(.
github said "Something went terribly wrong" when I tried uploading files ;).

1. The reason for the undef error is what I tried to explain above but obviously failed - let me try again :). The original test code makes use of an uninitialized value. However, if you don't have lifetime intrinsics that limit the scope of the struct to the loop body, the uninitialized value only exists in a loop phi node as incoming value from the preheader since the compiler doesn't know anymore that the struct definition is local and can never span multiple iterations. This is why it didn't manifests as an add with undef as one operand previously but now, with correct lifetime intrinsics, it does. 
2. I added the additional checks to ensure that there is a loop in the code. I'll gladly remove this again if you don't want this to be tested, it's indeed unrelated to the lifetime intrinsics.

Reply:
Right, so my concern wasn't with the undef part exactly, that part makes perfect sense.  I was just concerned whether this test case would still repro the original issue that the test was meant to repro, without the memcpy change that went along with it.
However, It sounds like it may no longer be a good test case for this memcpy change anymore due to other recent changes made to SROA, if I'm interpreting this right.  So it may be testing some special case, or maybe just a more common one now.  I can let it go if Greg can :).

Reply:
Just to clarify, I was purely speculating that the recent SROA changes might have rendered this test moot. I'm about 10% sure that's the case.

More relevant, I'm about 90% (independent of the previous 10%) sure that moving this definition doesn't negatively impact the intent of this test.